server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java

```diff
@@ -0,0 +1,472 @@
+import org.apache.lucene.index.Fields;
+import org.apache.lucene.index.FilterLeafReader;
+import org.apache.lucene.index.ImpactsEnum;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.index.TermState;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.RandomAccessInput;
+import org.apache.lucene.util.AttributeSource;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.lucene.store.IndexOutputOutputStream;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.ByteArray;
+import org.elasticsearch.core.IOUtils;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * This implementation is forked from Lucene's BloomFilterPosting to support on-disk bloom filters.
+ * <p>
+ * A {@link PostingsFormat} useful for low doc-frequency fields such as primary keys. Bloom filters
+ * offers "fast-fail" for reads in segments known to have no record of the key.
+ */
+public class BloomFilterPostingsFormat extends PostingsFormat {
+    static final String BLOOM_CODEC_NAME = "ESBloomFilter";
+    static final int VERSION_START = 1;
+    static final int VERSION_CURRENT = VERSION_START;
+    static final String BLOOM_FILTER_META_FILE = "bfm";
+    static final String BLOOM_FILTER_INDEX_FILE = "bfi";
+
+    private PostingsFormat postingsFormat;
+    private BigArrays bigArrays;
+
+    public BloomFilterPostingsFormat(PostingsFormat postingsFormat, BigArrays bigArrays) {
+        this();
+        this.postingsFormat = Objects.requireNonNull(postingsFormat);
+        this.bigArrays = Objects.requireNonNull(bigArrays);
+        if (postingsFormat instanceof PerFieldPostingsFormat) {
+            assert false : "Can't use PerFieldPostingsFormat with ESBloomFilterPostingsFormat";
+            throw new UnsupportedOperationException("Can't use PerFieldPostingsFormat with ESBloomFilterPostingsFormat");
+        }
+    }
+
+    public BloomFilterPostingsFormat() {
+        super(BLOOM_CODEC_NAME);
+    }
+
+    @Override
+    public FieldsConsumer fieldsConsumer(SegmentWriteState state) throws IOException {
+        if (postingsFormat == null || bigArrays == null) {
+            assert false : "ESBloomFilterPostingsFormat was initialized with a wrong constructor";
```

Comment:
nit: the message says `ESBloomFilterPostingsFormat` while the class is called `BloomFilterPostingsFormat`

Reply:
Should we rename this class to `ESBloomFilterPostingsFormat` instead?

Reply:
I renamed BloomFilterPostingsFormat to [ES84BloomFilterPostingsFormat](https://github.com/elastic/elasticsearch/pull/88409/commits/0fabf37e1973f9dedb4aff01b746c5341ea0c06f)